File: src/main/java/com/flexpoker/game/command/aggregate/Game.java
package com.flexpoker.game.command.aggregate;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import com.flexpoker.exception.FlexPokerException;
import com.flexpoker.framework.command.EventApplier;
import com.flexpoker.framework.domain.AggregateRoot;
import com.flexpoker.game.command.events.BlindsIncreasedEvent;
import com.flexpoker.game.command.events.GameCreatedEvent;
import com.flexpoker.game.command.events.GameFinishedEvent;
import com.flexpoker.game.command.events.GameJoinedEvent;
import com.flexpoker.game.command.events.GameMovedToStartingStageEvent;
import com.flexpoker.game.command.events.GameStartedEvent;
import com.flexpoker.game.command.events.GameTablesCreatedAndPlayersAssociatedEvent;
import com.flexpoker.game.command.events.NewHandIsClearedToStartEvent;
import com.flexpoker.game.command.events.PlayerBustedGameEvent;
import com.flexpoker.game.command.events.PlayerMovedToNewTableEvent;
import com.flexpoker.game.command.events.TablePausedForBalancingEvent;
import com.flexpoker.game.command.events.TableRemovedEvent;
import com.flexpoker.game.command.events.TableResumedAfterBalancingEvent;
import com.flexpoker.game.command.framework.GameEvent;
import com.flexpoker.game.query.dto.GameStage;

public class Game extends AggregateRoot<GameEvent> {

    private final UUID aggregateId;

    private int aggregateVersion;

    private final Map<Class<? extends GameEvent>, EventApplier<? super GameEvent>> methodTable;

    private GameStage gameStage;

    private final Set<UUID> registeredPlayerIds;

    private final int maxNumberOfPlayers;

    private final int numberOfPlayersPerTable;

    private final Map<UUID, Set<UUID>> tableIdToPlayerIdsMap;

    private final BlindSchedule blindSchedule;

    private final TableBalancer tableBalancer;

    private final Set<UUID> pausedTablesForBalancing;

    protected Game(boolean creatingFromEvents, UUID aggregateId,
            String gameName, int maxNumberOfPlayers,
            int numberOfPlayersPerTable, int numberOfSecondsForActionOnTimer,
            UUID createdById, GameStage gameStage, BlindSchedule blindSchedule,
            TableBalancer tableBalancer) {
        this.aggregateId = aggregateId;
        this.maxNumberOfPlayers = maxNumberOfPlayers;
        this.numberOfPlayersPerTable = numberOfPlayersPerTable;
        this.blindSchedule = blindSchedule;
        this.tableBalancer = tableBalancer;
        this.gameStage = gameStage;
        registeredPlayerIds = new HashSet<>();
        tableIdToPlayerIdsMap = new HashMap<>();
        pausedTablesForBalancing = new HashSet<>();

        methodTable = new HashMap<>();
        populateMethodTable();

        if (!creatingFromEvents) {
            var gameCreatedEvent = new GameCreatedEvent(
                    aggregateId, ++aggregateVersion, gameName,
                    maxNumberOfPlayers, numberOfPlayersPerTable, createdById,
                    blindSchedule.getNumberOfMinutesBetweenLevels(),
                    numberOfSecondsForActionOnTimer);
            addNewEvent(gameCreatedEvent);
            applyCommonEvent(gameCreatedEvent);
        }
    }

    @Override
    public void applyAllHistoricalEvents(List<GameEvent> events) {
        for (GameEvent event : events) {
            aggregateVersion++;
            applyCommonEvent(event);
        }
    }

    private void populateMethodTable() {
        methodTable.put(GameCreatedEvent.class, x -> {});
        methodTable.put(GameJoinedEvent.class, x -> registeredPlayerIds
                .add(((GameJoinedEvent) x).getPlayerId()));
        methodTable.put(GameMovedToStartingStageEvent.class, x -> gameStage = GameStage.STARTING);
        methodTable.put(GameStartedEvent.class, x -> gameStage = GameStage.INPROGRESS);
        methodTable.put(GameTablesCreatedAndPlayersAssociatedEvent.class,
                x -> tableIdToPlayerIdsMap
                        .putAll(((GameTablesCreatedAndPlayersAssociatedEvent) x)
                                .getTableIdToPlayerIdsMap()));
        methodTable.put(GameFinishedEvent.class, x -> gameStage = GameStage.FINISHED);
        methodTable.put(NewHandIsClearedToStartEvent.class, x -> {});
        methodTable.put(BlindsIncreasedEvent.class, x -> blindSchedule.incrementLevel());
        methodTable.put(TableRemovedEvent.class, x -> {
            var tableId = ((TableRemovedEvent) x).getTableId();
            tableIdToPlayerIdsMap.remove(tableId);
            pausedTablesForBalancing.remove(tableId);
        });
        methodTable.put(TablePausedForBalancingEvent.class, x -> {
            var tableId = ((TablePausedForBalancingEvent) x).getTableId();
            pausedTablesForBalancing.add(tableId);
        });
        methodTable.put(TableResumedAfterBalancingEvent.class, x -> {
            var tableId = ((TableResumedAfterBalancingEvent) x).getTableId();
            pausedTablesForBalancing.remove(tableId);
        });
        methodTable.put(PlayerMovedToNewTableEvent.class, x -> {
            var event = (PlayerMovedToNewTableEvent) x;
            tableIdToPlayerIdsMap.get(event.getFromTableId()).remove(event.getPlayerId());
            tableIdToPlayerIdsMap.get(event.getToTableId()).add(event.getPlayerId());
        });
        methodTable.put(PlayerBustedGameEvent.class, x -> {
            var event = (PlayerBustedGameEvent) x;
            var tableId = tableIdToPlayerIdsMap.entrySet().stream()
                    .filter(y -> y.getValue().contains(event.getPlayerId()))
                    .findAny().get().getKey();
            tableIdToPlayerIdsMap.get(tableId).remove(event.getPlayerId());
        });
    }

    private void applyCommonEvent(GameEvent event) {
        methodTable.get(event.getClass()).applyEvent(event);
        addAppliedEvent(event);
    }

    public void joinGame(UUID playerId) {
        createJoinGameEvent(playerId);

        // if the game is at the max capacity, start the game
        if (registeredPlayerIds.size() == maxNumberOfPlayers) {
            createGameMovedToStartingStageEvent();
            createGameTablesCreatedAndPlayersAssociatedEvent();
            createGameStartedEvent();
        }
    }

    public void attemptToStartNewHand(UUID tableId, Map<UUID, Integer> playerToChipsAtTableMap) {
        if (gameStage != GameStage.INPROGRESS) {
            throw new FlexPokerException("the game must be INPROGRESS if trying"
                    + "to start a new hand");
        }

        playerToChipsAtTableMap.entrySet().stream()
            .filter(x -> x.getValue() == 0)
            .map(x -> x.getKey())
            .forEach(x -> bustPlayer(x));

        var bustedPlayers = tableIdToPlayerIdsMap.get(tableId).stream()
                .filter(x -> !playerToChipsAtTableMap.keySet().contains(x))
                .collect(Collectors.toSet());
        bustedPlayers.forEach(x -> bustPlayer(x));

        if (tableIdToPlayerIdsMap.values().stream().flatMap(Collection::stream).count() == 1) {
            // TODO: do something for the winner
        } else {
            Optional<GameEvent> singleBalancingEvent;
            do {
                singleBalancingEvent = tableBalancer.createSingleBalancingEvent(aggregateVersion + 1,
                        tableId, pausedTablesForBalancing, tableIdToPlayerIdsMap, playerToChipsAtTableMap);
                if (singleBalancingEvent.isPresent()) {
                    aggregateVersion++;
                    addNewEvent(singleBalancingEvent.get());
                    applyCommonEvent(singleBalancingEvent.get());
                }
            } while (singleBalancingEvent.isPresent());

            if (tableIdToPlayerIdsMap.containsKey(tableId) && !pausedTablesForBalancing.contains(tableId)) {
                var event = new NewHandIsClearedToStartEvent(
                        aggregateId, ++aggregateVersion, tableId,
                        blindSchedule.getCurrentBlindAmounts());
                addNewEvent(event);
                applyCommonEvent(event);
            }
        }
    }

    public void increaseBlinds() {
        if (gameStage != GameStage.INPROGRESS) {
            throw new FlexPokerException(
                    "cannot increase blinds if the game isn't in progress");
        }

        if (!blindSchedule.isMaxLevel()) {
            var event = new BlindsIncreasedEvent(aggregateId, ++aggregateVersion);
            addNewEvent(event);
            applyCommonEvent(event);
        }
    }

    private void bustPlayer(UUID playerId) {
        if (tableIdToPlayerIdsMap.values().stream()
                .flatMap(Collection::stream)
                .noneMatch(x -> x.equals(playerId))) {
            throw new FlexPokerException("player is not active in the game");
        }

        var event = new PlayerBustedGameEvent(aggregateId, ++aggregateVersion, playerId);
        addNewEvent(event);
        applyCommonEvent(event);
    }

    private void createJoinGameEvent(UUID playerId) {
        if (gameStage == GameStage.STARTING || gameStage == GameStage.INPROGRESS) {
            throw new FlexPokerException("The game has already started");
        }

        if (gameStage == GameStage.FINISHED) {
            throw new FlexPokerException("The game is already finished.");
        }

        if (registeredPlayerIds.contains(playerId)) {
            throw new FlexPokerException("You are already in this game.");
        }

        var event = new GameJoinedEvent(aggregateId, ++aggregateVersion, playerId);
        addNewEvent(event);
        applyCommonEvent(event);
    }

    private void createGameMovedToStartingStageEvent() {
        if (gameStage != GameStage.REGISTERING) {
            throw new FlexPokerException(
                    "to move to STARTING, the game stage must be REGISTERING");
        }
        var event = new GameMovedToStartingStageEvent(aggregateId, ++aggregateVersion);
        addNewEvent(event);
        applyCommonEvent(event);
    }

    private void createGameTablesCreatedAndPlayersAssociatedEvent() {
        if (!tableIdToPlayerIdsMap.isEmpty()) {
            throw new FlexPokerException(
                    "tableToPlayerIdsMap should be empty when initializing the tables");
        }
        var tableIdToPlayerIdsMap = createTableToPlayerMap();

        var event = new GameTablesCreatedAndPlayersAssociatedEvent(
                aggregateId, ++aggregateVersion, tableIdToPlayerIdsMap,
                numberOfPlayersPerTable);
        addNewEvent(event);
        applyCommonEvent(event);
    }

    private Map<UUID, Set<UUID>> createTableToPlayerMap() {
        var randomizedListOfPlayerIds = new ArrayList<>(registeredPlayerIds);
        Collections.shuffle(randomizedListOfPlayerIds);

        var tableIdToPlayerIdsMap = new HashMap<UUID, Set<UUID>>();

        var numberOfTablesToCreate = determineNumberOfTablesToCreate();
        Stream.iterate(0, e -> e)
                .limit(numberOfTablesToCreate)
                .forEach(x -> tableIdToPlayerIdsMap.put(UUID.randomUUID(), new HashSet<>()));

        var tableIdList = new ArrayList<>(tableIdToPlayerIdsMap.keySet());

        Stream.iterate(0, e -> e + 1)
                .limit(randomizedListOfPlayerIds.size())
                .forEach(x -> {
                    var tableIndex = x % tableIdList.size();
                    tableIdToPlayerIdsMap.get(tableIdList.get(tableIndex))
                            .add(randomizedListOfPlayerIds.get(x));
                });

        return tableIdToPlayerIdsMap;
    }

    private int determineNumberOfTablesToCreate() {
        var numberOfTables = registeredPlayerIds.size() / numberOfPlayersPerTable;

        // if the number of people doesn't fit perfectly, then an additional
        // table is needed for the overflow
        if (registeredPlayerIds.size() % numberOfPlayersPerTable != 0) {
            numberOfTables++;
        }

        return numberOfTables;
    }

    private void createGameStartedEvent() {
        if (gameStage != GameStage.STARTING) {
            throw new FlexPokerException(
                    "to move to STARTED, the game stage must be STARTING");
        }

        if (tableIdToPlayerIdsMap.isEmpty()) {
            throw new FlexPokerException(
                    "tableToPlayerIdsMap should be filled at this point");
        }

        var event = new GameStartedEvent(aggregateId, ++aggregateVersion, tableIdToPlayerIdsMap.keySet(), blindSchedule);
        addNewEvent(event);
        applyCommonEvent(event);
    }

}

File: src/main/kotlin/com/flexpoker/game/command/aggregate/Game.kt
package com.flexpoker.game.command.aggregate

import com.flexpoker.exception.FlexPokerException
import com.flexpoker.game.command.events.BlindsIncreasedEvent
import com.flexpoker.game.command.events.GameCreatedEvent
import com.flexpoker.game.command.events.GameEvent
import com.flexpoker.game.command.events.GameFinishedEvent
import com.flexpoker.game.command.events.GameJoinedEvent
import com.flexpoker.game.command.events.GameMovedToStartingStageEvent
import com.flexpoker.game.command.events.GameStartedEvent
import com.flexpoker.game.command.events.GameTablesCreatedAndPlayersAssociatedEvent
import com.flexpoker.game.command.events.NewHandIsClearedToStartEvent
import com.flexpoker.game.command.events.PlayerBustedGameEvent
import com.flexpoker.game.command.events.PlayerMovedToNewTableEvent
import com.flexpoker.game.command.events.TablePausedForBalancingEvent
import com.flexpoker.game.command.events.TableRemovedEvent
import com.flexpoker.game.command.events.TableResumedAfterBalancingEvent
import com.flexpoker.game.query.dto.GameStage
import org.pcollections.HashTreePMap
import org.pcollections.HashTreePSet
import org.pcollections.PMap
import org.pcollections.PSet
import java.util.ArrayList
import java.util.HashMap
import java.util.Optional
import java.util.UUID

class Game constructor(
    creatingFromEvents: Boolean,
    aggregateId: UUID,
    gameName: String,
    maxNumberOfPlayers: Int,
    numberOfPlayersPerTable: Int,
    numberOfSecondsForActionOnTimer: Int,
    createdById: UUID,
    numberOfMinutesBetweenBlindLevels: Int
) {
    private var state = GameState(
        aggregateId,
        maxNumberOfPlayers,
        numberOfPlayersPerTable,
        GameStage.REGISTERING,
        blindSchedule(numberOfMinutesBetweenBlindLevels)
    )

    private val newEvents: MutableList<GameEvent>
    private val appliedEvents: MutableList<GameEvent>

    fun fetchNewEvents(): List<GameEvent> {
        return ArrayList(newEvents)
    }

    fun fetchAppliedEvents(): List<GameEvent> {
        return ArrayList(appliedEvents)
    }

    fun applyAllHistoricalEvents(events: List<GameEvent>) {
        events.forEach { applyCommonEvent(it) }
    }

    private fun applyCommonEvent(event: GameEvent) {
        when(event) {
            is GameCreatedEvent -> { }
            is GameJoinedEvent ->
                state = state.copy(registeredPlayerIds = state.registeredPlayerIds.plus(event.playerId))
            is GameMovedToStartingStageEvent -> state = state.copy(stage = GameStage.STARTING)
            is GameStartedEvent -> state = state.copy(stage = GameStage.INPROGRESS)
            is GameTablesCreatedAndPlayersAssociatedEvent ->
                state = state.copy(tableIdToPlayerIdsMap = state.tableIdToPlayerIdsMap.plusAll(event.tableIdToPlayerIdsMap))
            is GameFinishedEvent -> state = state.copy(stage = GameStage.FINISHED)
            is NewHandIsClearedToStartEvent -> { }
            is BlindsIncreasedEvent -> state = incrementLevel(state)
            is TableRemovedEvent -> {
                state = state.copy(tableIdToPlayerIdsMap = state.tableIdToPlayerIdsMap.minus(event.tableId))
                state = state.copy(pausedTablesForBalancing = state.pausedTablesForBalancing.minus(event.tableId))
            }
            is TablePausedForBalancingEvent ->
                state = state.copy(pausedTablesForBalancing = state.pausedTablesForBalancing.plus(event.tableId))
            is TableResumedAfterBalancingEvent ->
                state = state.copy(pausedTablesForBalancing = state.pausedTablesForBalancing.minus(event.tableId))
            is PlayerMovedToNewTableEvent -> {
                val fromPlayerIds = state.tableIdToPlayerIdsMap[event.fromTableId]!!.minus(event.playerId)
                val toPlayerIds = state.tableIdToPlayerIdsMap[event.fromTableId]!!.plus(event.playerId)
                state = state.copy(tableIdToPlayerIdsMap = state.tableIdToPlayerIdsMap.plus(event.fromTableId, fromPlayerIds))
                state = state.copy(tableIdToPlayerIdsMap = state.tableIdToPlayerIdsMap.plus(event.toTableId, toPlayerIds))
            }
            is PlayerBustedGameEvent -> {
                val tableId = state.tableIdToPlayerIdsMap.entries.first { it.value.contains(event.playerId) }.key
                val playerIds = state.tableIdToPlayerIdsMap[tableId]!!.minus(event.playerId)
                state = state.copy(tableIdToPlayerIdsMap = state.tableIdToPlayerIdsMap.plus(tableId, playerIds))
            }
        }
        appliedEvents.add(event)
    }

    fun joinGame(playerId: UUID) {
        createJoinGameEvent(playerId)

        // if the game is at the max capacity, start the game
        if (state.registeredPlayerIds.size == state.maxNumberOfPlayers) {
            createGameMovedToStartingStageEvent()
            createGameTablesCreatedAndPlayersAssociatedEvent()
            createGameStartedEvent()
        }
    }

    fun attemptToStartNewHand(tableId: UUID, playerToChipsAtTableMap: Map<UUID, Int>) {
        if (state.stage !== GameStage.INPROGRESS) {
            throw FlexPokerException("the game must be INPROGRESS if trying to start a new hand")
        }
        playerToChipsAtTableMap.filterValues { it == 0 }.map { it.key }.forEach { bustPlayer(it) }

        val bustedPlayers = state.tableIdToPlayerIdsMap[tableId]!!
            .filter { !playerToChipsAtTableMap.keys.contains(it) }.toSet()
        bustedPlayers.forEach { bustPlayer(it) }

        if (state.tableIdToPlayerIdsMap.flatMap { it.value }.count() == 1) {
            // TODO: do something for the winner
        } else {
            var singleBalancingEvent: Optional<GameEvent>?
            do {
                singleBalancingEvent = createSingleBalancingEvent(state.aggregateId, state.maxNumberOfPlayers, tableId,
                    state.pausedTablesForBalancing, state.tableIdToPlayerIdsMap, playerToChipsAtTableMap)
                if (singleBalancingEvent.isPresent) {
                    newEvents.add(singleBalancingEvent.get())
                    applyCommonEvent(singleBalancingEvent.get())
                }
            } while (singleBalancingEvent!!.isPresent)
            if (state.tableIdToPlayerIdsMap.containsKey(tableId) && !state.pausedTablesForBalancing.contains(tableId)) {
                val event = NewHandIsClearedToStartEvent(state.aggregateId, tableId, currentBlindAmounts(state))
                newEvents.add(event)
                applyCommonEvent(event)
            }
        }
    }

    fun increaseBlinds() {
        if (state.stage !== GameStage.INPROGRESS) {
            throw FlexPokerException("cannot increase blinds if the game isn't in progress")
        }
        if (!isMaxLevel(state)) {
            val event = BlindsIncreasedEvent(state.aggregateId)
            newEvents.add(event)
            applyCommonEvent(event)
        }
    }

    private fun bustPlayer(playerId: UUID) {
        if (state.tableIdToPlayerIdsMap.flatMap { it.value }.none { it == playerId }) {
            throw FlexPokerException("player is not active in the game")
        }
        val event = PlayerBustedGameEvent(state.aggregateId, playerId)
        newEvents.add(event)
        applyCommonEvent(event)
    }

    private fun createJoinGameEvent(playerId: UUID) {
        if (state.stage === GameStage.STARTING || state.stage === GameStage.INPROGRESS) {
            throw FlexPokerException("The game has already started")
        }
        if (state.stage === GameStage.FINISHED) {
            throw FlexPokerException("The game is already finished.")
        }
        if (state.registeredPlayerIds.contains(playerId)) {
            throw FlexPokerException("You are already in this game.")
        }
        val event = GameJoinedEvent(state.aggregateId, playerId)
        newEvents.add(event)
        applyCommonEvent(event)
    }

    private fun createGameMovedToStartingStageEvent() {
        if (state.stage !== GameStage.REGISTERING) {
            throw FlexPokerException("to move to STARTING, the game stage must be REGISTERING")
        }
        val event = GameMovedToStartingStageEvent(state.aggregateId)
        newEvents.add(event)
        applyCommonEvent(event)
    }

    private fun createGameTablesCreatedAndPlayersAssociatedEvent() {
        if (state.tableIdToPlayerIdsMap.isNotEmpty()) {
            throw FlexPokerException("tableToPlayerIdsMap should be empty when initializing the tables")
        }
        val tableMap = createTableToPlayerMap()
        val event = GameTablesCreatedAndPlayersAssociatedEvent(state.aggregateId, tableMap, state.numberOfPlayersPerTable)
        newEvents.add(event)
        applyCommonEvent(event)
    }

    private fun createTableToPlayerMap(): PMap<UUID, PSet<UUID>> {
        val randomizedListOfPlayerIds = state.registeredPlayerIds.toList().shuffled()
        val numberOfTablesToCreate = determineNumberOfTablesToCreate()

        val tableMap = (0 until numberOfTablesToCreate)
            .fold(HashMap<UUID, PSet<UUID>>(), { acc, _ ->
                acc[UUID.randomUUID()] = HashTreePSet.empty()
                acc
            })

        val tableIdList = ArrayList(tableMap.keys)
        randomizedListOfPlayerIds.indices.forEach {
            val tableIndex = it % tableIdList.size
            val players = tableMap[tableIdList[tableIndex]]!!.plus(randomizedListOfPlayerIds[it])
            tableMap[tableIdList[tableIndex]] = players
        }

        return HashTreePMap.from(tableMap)
    }

    private fun determineNumberOfTablesToCreate(): Int {
        var numberOfTables = state.registeredPlayerIds.size / state.numberOfPlayersPerTable

        // if the number of people doesn't fit perfectly, then an additional
        // table is needed for the overflow
        if (state.registeredPlayerIds.size % state.numberOfPlayersPerTable != 0) {
            numberOfTables++
        }
        return numberOfTables
    }

    private fun createGameStartedEvent() {
        if (state.stage !== GameStage.STARTING) {
            throw FlexPokerException("to move to STARTED, the game stage must be STARTING")
        }
        if (state.tableIdToPlayerIdsMap.isEmpty()) {
            throw FlexPokerException("tableToPlayerIdsMap should be filled at this point")
        }
        val event = GameStartedEvent(state.aggregateId, state.tableIdToPlayerIdsMap.keys, state.blindScheduleDTO)
        newEvents.add(event)
        applyCommonEvent(event)
    }

    init {
        newEvents = ArrayList()
        appliedEvents = ArrayList()

        if (!creatingFromEvents) {
            val gameCreatedEvent = GameCreatedEvent(
                aggregateId, gameName, maxNumberOfPlayers,
                state.numberOfPlayersPerTable, createdById, numberOfMinutesBetweenLevels(state),
                numberOfSecondsForActionOnTimer
            )
            newEvents.add(gameCreatedEvent)
            applyCommonEvent(gameCreatedEvent)
        }
    }
}